Sephora-M/assignment2/BinaryTreeTest.java

```diff
@@ -0,0 +1,104 @@
+package assignment2;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.Rule;
+import org.junit.jupiter.api.Test;
+import org.junit.rules.ExpectedException;
+import java.util.NoSuchElementException;
+
+class BinaryTreeTest {
+	
+	@Rule
+    public ExpectedException thrown = ExpectedException.none();
+	
+	@Test
+	void testContainsNode() {
+		BinaryTreeNode<String> a = new BinaryTreeNode<String>("a", null);
+		BinaryTree<String> A = new BinaryTree<String>(a);
+		
+		A.add("b");
+		A.add("c");
+		A.add("d");
+		A.add("e");
+		A.add("f");
+		A.add("g");
+		A.add("h");
+		A.add("i");
+		
+		assertTrue(A.find("h") != null);
+		assertTrue(A.find("c") != null);
+		assertTrue(A.find("f") != null);
+		assertFalse(A.find("j") != null);
+	}
+	
+	@Test
+	void testAddChild() {
+		BinaryTreeNode<String> a = new BinaryTreeNode<String>("a", null);
+		a.addChild("b");
+		a.addChild("c");
+		
+		BinaryTreeNode<String> b = a.findNode("b");
+		
+		assertTrue(b.getParent() == a);
+	}
+	
+	@Test
+	void testFindNode(){
+		BinaryTreeNode<String> a = new BinaryTreeNode<String>("a", null);
+		a.addChild("b");
+		a.addChild("c");
+		BinaryTree<String> A = new BinaryTree<String>(a);
+	
+		try {
+			A.printAncestors("d");
+			assert false;
+		} catch (NoSuchElementException e) {
+			assert true;
+		}
+	}
+	
+	@Test
+	void testCommonAncestor() {
+		BinaryTreeNode<String> a = new BinaryTreeNode<String>("a", null);
+		BinaryTreeNode<String> b = new BinaryTreeNode<String>("b", null);
+		BinaryTreeNode<String> c = new BinaryTreeNode<String>("c", null);
+		BinaryTreeNode<String> d = new BinaryTreeNode<String>("d", null);
+		BinaryTreeNode<String> e = new BinaryTreeNode<String>("e", null);
+		BinaryTreeNode<String> f = new BinaryTreeNode<String>("f", null);
+		BinaryTreeNode<String> g = new BinaryTreeNode<String>("g", null);
+		BinaryTreeNode<String> h = new BinaryTreeNode<String>("h", null);
+		BinaryTreeNode<String> i = new BinaryTreeNode<String>("i", null);
+		BinaryTreeNode<String> j = new BinaryTreeNode<String>("j", null);
+		BinaryTreeNode<String> k = new BinaryTreeNode<String>("k", null);
+		BinaryTreeNode<String> l = new BinaryTreeNode<String>("l", null);
+		BinaryTreeNode<String> m = new BinaryTreeNode<String>("m", null);
+		BinaryTreeNode<String> n = new BinaryTreeNode<String>("n", null);
+		
+		a.addChild(b);
```

Comment:
optional (but you may often see it at interviews):
It looks like you have two methods for creating the tree: 
a) putting a child at a random leaf of the tree
b) using the addChild at the specified node. 

Do you think you could add another method, that would take an inorder/postorder/preorder description of a tree? You could then also add a method that prints a inorder/postorder/preoder description of a tree to verify that the construction is valid. You can use brackets "(" and ")" for example to illustrate nodes and parents.

Reply:
Yes I added the method, please have a look 😇